cmd/argocd/commands/login.go

```diff
@@ -38,35 +47,48 @@ func NewLoginCommand(globalClientOpts *argocdclient.ClientOptions) *cobra.Comman
 			errors.CheckError(err)
 			if !tlsTestResult.TLS {
 				if !globalClientOpts.PlainText {
-					askToProceed("WARNING: server is not configured with TLS. Proceed (y/n)? ")
+					cli.AskToProceed("WARNING: server is not configured with TLS. Proceed (y/n)? ")
 					globalClientOpts.PlainText = true
 				}
 			} else if tlsTestResult.InsecureErr != nil {
 				if !globalClientOpts.Insecure {
-					askToProceed(fmt.Sprintf("WARNING: server certificate had error: %s. Proceed insecurely (y/n)? ", tlsTestResult.InsecureErr))
+					cli.AskToProceed(fmt.Sprintf("WARNING: server certificate had error: %s. Proceed insecurely (y/n)? ", tlsTestResult.InsecureErr))
 					globalClientOpts.Insecure = true
 				}
 			}
-
-			username, password = cli.PromptCredentials(username, password)
-
 			clientOpts := argocdclient.ClientOptions{
 				ConfigPath: "",
 				ServerAddr: server,
 				Insecure:   globalClientOpts.Insecure,
 				PlainText:  globalClientOpts.PlainText,
 			}
-			conn, sessionIf := argocdclient.NewClientOrDie(&clientOpts).NewSessionClientOrDie()
-			defer util.Close(conn)
+			acdClient := argocdclient.NewClientOrDie(&clientOpts)
+			setConn, setIf := acdClient.NewSettingsClientOrDie()
+			defer util.Close(setConn)
+
+			acdSet, err := setIf.Get(context.Background(), &settings.SettingsQuery{})
+			errors.CheckError(err)
+
+			ctxName = cli.PromptMessage("Enter a name for this context", ctxName)
+			username = cli.PromptUsername(username)
 
-			sessionRequest := session.SessionCreateRequest{
-				Username: username,
-				Password: password,
+			// Perform the login
+			var tokenString string
+			if username == "admin" || !ssoConfigured(acdSet) {
+				tokenString = passwordLogin(acdClient, username, password)
+			} else {
+				tokenString = oauth2Login(server, clientOpts.PlainText)
+				tokenString = tokenLogin(acdClient, tokenString)
 			}
-			createdSession, err := sessionIf.Create(context.Background(), &sessionRequest)
+
+			parser := &jwt.Parser{
+				SkipClaimsValidation: true,
+			}
+			claims := jwt.MapClaims{}
+			_, _, err = parser.ParseUnverified(tokenString, &claims)
```

Comment:
Couple of reasons:
1. The only reason for parsing the claims at all, is for presentation purposes. In the CLI, I want to print identity of the user who just logged in, specifically the e-mail address field (if available). In this case, we are not parsing claims for presentation not authentication.

2. There are only two ways an ArgoCD token can get verified, depending on the issuer. If the issuer was from dex, we need to make a few extra calls to the Dex IDP to get his public key to verify it was truly signed by him. If the issuer was ArgoCD, it is impossible to do verification on the client, because currently ArgoCD does not expose a public key for verification purposes. None of these really add much to security because in the end, the server will be the one doing verification. So in the worst case, we would have stored an invalid token in the home directory of the user.

Reply:
@jessesuen thank you kindly for explaining these points to me.  My understanding was that since the server holds a signing key, the server is capable of validating the signature on the JWT tokens using HMAC.  Without validation, I'm concerned about the potential for token tampering, as I had thought that token/claims validation would ensure that not only is the token valid and signed by the server (or by Dex, which I think could be bolted on), but that nothing has been manipulated in any of the fields.  Is that an orthogonal concern here?

Reply:
@jessesuen I think we're on the same page regarding how easy it is to modify a token.  The question I have is whether disabling the specific checks I commented on here makes it possible to do such a manipulation successfully, because I thought said checks were specifically intended to make such tampering harder.

AAaaand okay, I see this is the command-line tools... not the server-side components.  Right.  Never mind!